firedrake/assemble.py

```diff
@@ -146,9 +150,17 @@ def _assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
               appctx={},
               collect_loops=False,
               allocate_only=False):
-    """Assemble the form f and return a Firedrake object representing the
-    result. This will be a :class:`float` for 0-forms, a
-    :class:`.Function` for 1-forms and a :class:`.Matrix` for 2-forms.
+    """Assemble the form or Slate expression f and return a Firedrake object
+    representing the result. This will be a :class:`float` for 0-forms, a
+    :class:`.Function` for 1-forms/rank-1 Slate tensors and a :class:`.Matrix`
+    for 2-forms/rank-2 Slate tensors.
+
+    .. warning::
+
+    If f is a 0-form wrapped as a Slate tensor, this will fail. 0-form tensors
+    cannot be expressed as an `Eigen::MatrixBase` object. However, rank-0
+    tensors may be used in a Slate expression, so long as the resulting rank
+    of the expression is 1 or 2.
```

Comment:
I saw that the deep in the compiler you raise a `NotImplementedError` in the rank-0 case.  I think it's friendlier to someone reading the code if the argument checking happens high up.